Garland 2P solution (#1009)
cpinitiative/usaco-guide: solutions/CF Garland.mdx

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang
---

[Official DP Editorial](https://codeforces.com/blog/entry/52449)

## Solution (Two Pointer)

**Time Complexity:** $\mathcal{O}(nq)$

> We can keep a sliding window of maximum consecutive runs of that letter with a maximum of $m$ letters exchanged.

Reviewer: Probably should elaborate.

Reviewer: Also sliding window or running sum?

Reviewer: Woops nvm

Reviewer 2: For each query we can keep a sliding window where no more than $a$ letters in the window are not $b$

something more like this???
idk

Reviewer 2: oh the variables in your soln are different from those in the problem

<LanguageSection>

<CPPSection>

### C++ Implementation:

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;
using ll = long long;
using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair
void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}

int N, T;
string W;

int main() {
	setIO();
	cin >> N >> W >> T;

	for(int i = 0; i < T; i++){
		int a, m = 0; char b;
		cin >> a >> b;
		
		int l, r;
		l = r = 0;
		while(l < N && r < N) {
			while(r < N) {
				if(W[r] != b) {
					if(a == 0) break;
					a--;
				}
				r++;
			}

			m = max(m, r - l);

> 			a += W[l++]==b?0:1;

Reviewer: ```suggestion
			a += W[l++] == b ? 0 : 1;
```

> 			a += W[l++] == b ? 0 : 1;

Reviewer: ```suggestion
			a += !W[l++] == b
```
could also do this ^ if you want :p

Reviewer 2: ! comes before == in operator precedence

Reviewer 3: Ehhh idt this matters lol

Reviewer 2: think it's fine as is tho

Reviewer: yeah me too

> 			a += W[l++] == b ? 0 : 1;

Reviewer: er not totally sure on the parantheses maybe 
```suggestion
			a += !(W[l++] == b);
```

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang, Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/52449)

## Solution (Two Pointers)

**Time Complexity:** $\mathcal{O}(nq)$

For each query we can keep a sliding window where no more than $m$ letters in the window are not $c$.

### Implementation

<LanguageSection>

<CPPSection>

```cpp
CodeSnip{TemplateShort}

int N, Q;
string S;

int main() {
	setIO();
	cin >> N >> S >> Q;

	for(int i = 0; i < Q; i++){
		int m, ma = 0; char c;
		cin >> m >> c;
		
		int l, r;
		l = r = 0;
		while(l < N && r < N) {
			while(r < N) {
				if(S[r] != c) {
					if(m == 0) break;
					m--;
				}
				r++;
			}

			ma = max(ma, r - l);

			m += S[l++] != c;
		}

		cout << ma << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Two Pointers with Prefix Sums)

**Time Complexity:** $\mathcal{O}(nq)$

> For each query $q$ we can use prefix sums to check the number of occurrences of a given letter $c$ and use two pointers to find the maximum Koyomity ending at a given index $j$ from $0 \dots n$.

Reviewer 4: > given index $j$ from $0 \dots n$.

not sure what j refers to in the code below

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;

int dp[26][1501];

int main(){
	int n; cin >> n;
	string s; cin >> s;
	for(int i=1; i<=n; i++){
		int color = s[i-1] - 'a';
		dp[color][i]++;
	}
	for(int i=0; i<26; i++){
		for(int j=1; j<=n; j++){
			dp[i][j] = dp[i][j-1] + dp[i][j];
		}
	}
	int q; cin >> q;
	for(int i=0; i<q; i++){
		int m; char c; cin >> m >> c;
		c -= 'a';
		int p1 = 1, p2 = 1, koyomity = 1;
		while(p2 <= n){
			int c_count = dp[c][p2] - dp[c][p1-1] + m;

> 			if((c_count) < (p2 - p1 + 1)){

Reviewer 4: these parentheses are not necessary, same below

				p1++;
			} else {
				koyomity = max(koyomity, (p2 - p1 + 1));
				p2++;
			}
		}
		cout << koyomity << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Dynamic Programming)

<Warning>

Dynamic programming is a [Gold](/gold/intro-dp) topic.

</Warning>

**Time Complexity:** $\mathcal{O}(n^2k + q)$

> We are performing up to $200\,000$ queries, so we must have some way to quickly precompute our answer in constant time.

Reviewer 4: > must

still not changed

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang, Jesse Choe
---

> [Official DP Editorial](https://codeforces.com/blog/entry/52449)

Reviewer 4: how is this DP?

Author: wait thats mb, i thought nwatx put a dp editorial.

## Solution (Two Pointers)

**Time Complexity:** $\mathcal{O}(nq)$

> For each query we can keep a sliding window where no more than $a$ letters in the window are not $b$

Reviewer 4: no period

				if(W[r] != b) {
					if(a == 0) break;
					a--;
				}
				r++;
			}

			m = max(m, r - l);

			a += W[l++] != b;
		}

		cout << m << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Two Pointers with Prefix Sums)

**Time Complexity:** $\mathcal{O}(nq)$

For each query we can use prefix sums to check the number of occurrences of a given letter $a$ and use two pointers to find the maximum Koyomity ending at a given index $j$.

<LanguageSection>

<CPPSection>

### C++ Implementation:

```cpp
#include <bits/stdc++.h>

using namespace std;

int dp[26][1501];

int main(){
	int n; cin >> n;
	string s; cin >> s;
	for(int i=1; i<=n; i++){
		int color = s[i-1] - 'a';
		dp[color][i]++;
	}
	for(int i=0; i<26; i++){
		for(int j=1; j<=n; j++){
			dp[i][j] = dp[i][j-1] + dp[i][j];
		}
	}
	int q; cin >> q;
	for(int i=0; i<q; i++){
		int m; char c; cin >> m >> c;
		c -= 'a';
		int p1 = 1, p2 = 1, koyomity = 1;
		while(p2 <= n){
			int c_count = dp[c][p2] - dp[c][p1-1] + m;
			if((c_count) < (p2 - p1 + 1)){
				p1++;
			} else {
				koyomity = max(koyomity, (p2 - p1 + 1));
				p2++;
			}
		}
		cout << koyomity << endl;
	}
}
```

</CPPSection>

</LanguageSection>

> ## Alternative Solution (Dynamic Programming)

Reviewer 3: We should remove this solution tbh...

Author: why? just because dp isnt in silver doesn't mean we cant have dp sol???

Reviewer 3: > why? just because dp isnt in silver doesn't mean we cant have dp sol???

I mean idt we need it tbh... At least we should put a note saying that this solution is for Gold+ ppl.

Author: i thought u did put a note...

Reviewer 3: Yeah, I wanted to make sure you guys approve of it...

Author: also sols are for everyone

> **Note: you do not need to understand this solution at the Silver level.**

Reviewer 4: I'd prefer something like

```
<Warning>

Dynamic programming is a [Gold](/gold/intro-dp) topic.

</Warning>
```

**Time Complexity:** $\mathcal{O}(n^2k + q)$

> We are performing up to $200\,000$ queries, so we must have some way to quickly precompute our answer in constant time.

Reviewer 4: some sentences below also missing period

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang, Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/52449)

## Solution (Two Pointers)

**Time Complexity:** $\mathcal{O}(nq)$

> For each query we can keep a sliding window where no more than $a$ letters in the window are not $b$.

Reviewer 4: Use the same variable names as the problem statement.

> For each query we can keep a sliding window where no more than $a$ letters in the window are not $b$.

Reviewer 3: Try this...
```suggestion
For each query, we can keep a sliding window with starting at index $i$ and ending at index $j$, where exactly $j-i$ letters are $c_i$
```

<LanguageSection>

<CPPSection>

### C++ Implementation:

```cpp

> #include <bits/stdc++.h> // see /general/running-code-locally

Reviewer 4: replace this part w/ codesnip?

Reviewer 4: Shouldn't it be `CodeSnip{TemplateShort}` as described here?

https://github.com/cpinitiative/usaco-guide/blob/d03f6397709701cc57809068af7cf0af4009cfed/docs/Content%20Documentation.md#code-blocks-and-code-snippets

using namespace std;
using ll = long long;
using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair
void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}

int N, Q;
string W;

int main() {
	setIO();
	cin >> N >> W >> Q;

	for(int i = 0; i < Q; i++){

> 		int a, m = 0; char b;

Reviewer 4: same here

**Time Complexity:** $\mathcal{O}(nq)$

For each query $q$ we can use prefix sums to check the number of occurrences of a given letter $c_i$ and use two pointers to find the maximum Koyomity ending at a given index $j$ from $0 \dots n$.

<LanguageSection>

<CPPSection>

### C++ Implementation:

```cpp
#include <bits/stdc++.h>

using namespace std;

int dp[26][1501];

int main(){
	int n; cin >> n;
	string s; cin >> s;
	for(int i=1; i<=n; i++){
		int color = s[i-1] - 'a';
		dp[color][i]++;
	}
	for(int i=0; i<26; i++){
		for(int j=1; j<=n; j++){
			dp[i][j] = dp[i][j-1] + dp[i][j];
		}
	}
	int q; cin >> q;
	for(int i=0; i<q; i++){
		int m; char c; cin >> m >> c;
		c -= 'a';
		int p1 = 1, p2 = 1, koyomity = 1;
		while(p2 <= n){
			int c_count = dp[c][p2] - dp[c][p1-1] + m;
			if((c_count) < (p2 - p1 + 1)){
				p1++;
			} else {
				koyomity = max(koyomity, (p2 - p1 + 1));
				p2++;
			}
		}
		cout << koyomity << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Dynamic Programming)

<Warning>

Dynamic programming is a [Gold](/gold/intro-dp) topic.

</Warning>

**Time Complexity:** $\mathcal{O}(n^2k + q)$

> We are performing up to $200\,000$ queries, so we must have some way to quickly precompute our answer in constant time.

Reviewer 3: ```suggestion
We are performing up to $200\,000$ queries, so it may be optimal to have some way to quickly precompute our answer in constant time.
```

> We are performing up to $200\,000$ queries, so we must have some way to quickly precompute our answer in constant time.

Reviewer 4: "must"? - what about the solutions above?

Author: @nwatx :))

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang, Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/52449)

## Solution (Two Pointers)

**Time Complexity:** $\mathcal{O}(nq)$

For each query we can keep a sliding window where no more than $m$ letters in the window are not $c$.

<LanguageSection>

<CPPSection>

> ### C++ Implementation:

Reviewer 4: Why colon?

using namespace std;
using ll = long long;
using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair
void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}
//EndCodeSnip

int N, Q;
string S;

int main() {
	setIO();
	cin >> N >> S >> Q;

	for(int i = 0; i < Q; i++){
		int m, ma = 0; char c;
		cin >> m >> c;
		
		int l, r;
		l = r = 0;
		while(l < N && r < N) {
			while(r < N) {
				if(S[r] != c) {
					if(m == 0) break;
					m--;
				}
				r++;
			}

			ma = max(ma, r - l);

			m += S[l++] != c;
		}

		cout << ma << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Two Pointers with Prefix Sums)

**Time Complexity:** $\mathcal{O}(nq)$

For each query $q$ we can use prefix sums to check the number of occurrences of a given letter $c$ and use two pointers to find the maximum Koyomity ending at a given index $j$ from $0 \dots n$.

<LanguageSection>

<CPPSection>

> ### C++ Implementation:

Reviewer 4: why colon?

id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang, Qi Wang, Jesse Choe
---

[Official Editorial](https://codeforces.com/blog/entry/52449)

## Solution (Two Pointers)

**Time Complexity:** $\mathcal{O}(nq)$

For each query we can keep a sliding window where no more than $m$ letters in the window are not $c$.

### Implementation

<LanguageSection>

<CPPSection>

```cpp
CodeSnip{TemplateShort}

int N, Q;
string S;

int main() {
	setIO();
	cin >> N >> S >> Q;

	for(int i = 0; i < Q; i++){
		int m, ma = 0; char c;
		cin >> m >> c;
		
		int l, r;
		l = r = 0;
		while(l < N && r < N) {
			while(r < N) {
				if(S[r] != c) {
					if(m == 0) break;
					m--;
				}
				r++;
			}

			ma = max(ma, r - l);

			m += S[l++] != c;
		}

		cout << ma << endl;
	}
}
```

</CPPSection>

</LanguageSection>

## Alternative Solution (Two Pointers with Prefix Sums)

> **Time Complexity:** $\mathcal{O}(nq)$

Reviewer 2: ```suggestion
**Time Complexity:** $\mathcal{O}(n(k+q))$
```

Technically this is more accurate.

2. We append the character, but since it is the same as our previous character we get to "save" a paint.<center>$$\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k] + 1)$$</center>

Since this only gives us the best possible score that ends with length $i$, we can just take the current best and translate it using $\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k])$.

Now, we can query the max painting score for any substring containing the first character. However, this is not required; we only have to query the optimal paintings with size $n$ (the full string) with $\texttt{dp}[n][m_i][c_i]$.

<LanguageSection>

<CPPSection>

> ## C++ Implementation

Reviewer 4: Should be `###`

(though why don't we just have one `### Implementation` above the languagesection instead of a separate `### [Language] Implementation` for each language?)

Author: sure but we usually to [language] impl right?

Reviewer 4: Really? I see `### Implementation` used most of the time (though most of the time it's inside the LanguageSection, I think it would be better outside).

Author: ok

2. We append the character, but since it is the same as our previous character we get to "save" a paint.<center>$$\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k] + 1)$$</center>

Since this only gives us the best possible score that ends with length $i$, we can just take the current best and translate it using $\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k])$.

Now, we can query the max painting score for any substring containing the first character. However, this is not required; we only have to query the optimal paintings with size $n$ (the full string) with $\texttt{dp}[n][m_i][c_i]$.

<LanguageSection>

<CPPSection>

```cpp

> #include <bits/stdc++.h>

Reviewer 4: same here
